app/services/schedule_solver.rb

```diff
@@ -129,80 +139,45 @@ def self.solve(classes, rooms, times, instructors, locks)
       (0...num_rooms).each do |r|
         (0...num_times).each do |t|
           next unless sched[c][r][t].value
 
-          if before_9?(times[t])
-            morning_classes.append(c)
-          elsif after_3?(times[t])
-            evening_classes.append(c)
-          end
+          classes[c]['time_slot'] = times[t]
+          classes[c]['room_id'] = rooms[r]['id']
         end
       end
     end
 
-    # Fill in unhappiness matrix
-    morning_haters.each do |p|
-      morning_classes.each do |c|
-        unhappiness_matrix[hash[p]][c] = 1
-      end
-    end
-
-    evening_haters.each do |p|
-      evening_classes.each do |c|
-        unhappiness_matrix[hash[p]][c] = 1
-      end
-    end
+    # Generate unhappiness matrix with prof ID
+    unhappiness_matrix, prof_ids = generate_unhappiness_matrix(classes, instructors, hours)
 
     # Solve the min weight perfect matching via the Hungarian algorithm
-    # While extensions to nonperfect matchings exist for HA, this particular library doesn't support them
-    # The library modifies the matrix, so create a deep copy first
+    # The library clobbers the matrix, so create a deep copy first
     copy = unhappiness_matrix.map(&:dup)
     matching = HungarianAlgorithm.new(unhappiness_matrix).process.sort
 
     # matching is a 2D array, where each element [i,j] represents the assignment of professsor i to class j
-    # Flatten this to a simpler map of profs to classes
-    # i.e have matching[i] = j instead of matching[i] = [i,j]
-    matching = matching.map { |_u, v| v }
+    # Map the course to prof_ids[prof], which gives the position of the true professor in the instructors array
+    matching = matching.map { |prof, course| [classes[course], prof_ids[prof]] }.to_h
 
     total_unhappiness = 0
-    (0...num_classes).each do |c|
-      assigned_prof = instructors[matching.index(c)]
-      (0...num_rooms).each do |r|
-        (0...num_times).each do |t|
-          next unless sched[c][r][t].value
-
-          unhappiness = copy[hash[assigned_prof]][c]
-          total_unhappiness += unhappiness
-
-          RoomBooking.create(
-            room_id: rooms[r]['id'],
-            time_slot_id: times[t][3],
-            is_available: true,
-            is_lab: false,
-            created_at: Time.now,
-            updated_at: Time.now,
-            course_id: classes[c]['id'],
-            instructor_id: assigned_prof['id']
-          )
-        end
-      end
+    classes.each do |assigned_course|
+      true_prof_id = matching[assigned_course]
+      course_id = assigned_course['id']
+      total_unhappiness += copy[true_prof_id][class_id_hash[course_id]]
+      RoomBooking.create(
+        room_id: assigned_course['room_id'],
+        time_slot_id: assigned_course['time_slot'][3],
+        is_available: true,
+        is_lab: false,
+        created_at: Time.now,
+        updated_at: Time.now,
+        course_id:,
+        instructor_id: instructors[true_prof_id]['id']
```

Comment:
Could we also ensure that we block the overlapping time_slots here?

Reply:
Fixed, see 73d031b